oscarlab/graphene
# [Pal/Linux-SGX] Add TLS-PSK protection to process checkpoint IPC

## Affected components


## Description of the changes 

Previously, Graphene-SGX did not protect send/recv of checkpoint from parent to child. This leaked all memory contents of the parent process.

This PR adds TLS-PSK (TLS with Pre-Shared Key) wrapper for process communication. Graphene-SGX already has the logic for SGX-based local attestation and generation of the shared key for each parent <-> child communication channel via Diffie–Hellman key exchange. This PR uses this pre-shared key to create an mbedTLS-based session based on UNIX domain socketpair (`parent.process.stream` <-> `child.process.stream`).

`_DkStreamSecure{Init,Free,Read,Write}` internal Linux-SGX PAL functions are added and used during child process creation and sending of the parent-generated checkpoint. These functions are backed by crypto-layer `lib_SSL{Init,Free,Read,Write}` functions which in turn use mbedTLS. Configuration of mbedTLS is expanded to support TLS-PSK; note that for entropy source we use only rdrand instruction inside SGX enclave (i.e., no untrusted host-platform sources of entropy). The only ciphersuite currently supported for IPC is `MBEDTLS_TLS_PSK_WITH_AES_128_GCM_SHA256`.

This PR adds protection **only** for checkpoint IPC. After the checkpoint is sent/received, the parent/child processes downgrade their sockets to plain non-secure ones (by disabling TLS). This is done because child may spawn a grandchild that also wants to communicate with grandparent, but it is impossible for multiple processes to share same TLS context.

Additionally, this PR sanitizes "process" PAL handles during send-handle checkpoint send (via `DkSendHandle`): shared key and TLS context are zeroed out.

*Interesting references*:
- "Comparison studies between pre-shared and public key exchange mechanisms for Transport Layer Security" by Fang-Chun Kuo et al. Nice introduction to TLS-PSK and why it's needed.
- RFC 4279 and 5487. Official documents for TLS-PSK.
- https://security.stackexchange.com/questions/3204/computationally-simple-lightweight-replacement-for-ssl-tls. Basically: never develop your own crypto-schemes unless you're Rivest/Shamir/Adleman.
- https://www.wolfssl.com/when-to-use-pre-shared-key-psk-cipher-suites-2/.

## How to test this PR? 

All tests must pass.

If you test manually, note that you must completely re-build mbedTLS (it is not rebuilt by default). Do this before re-build in the main Graphene dir: `graphene/Pal/lib$ make distclean SGX=1`.

To test that checkpoint is actually sent encrypted with TLS-PSK, one can use `vfork_and_exec` LibOS test like this:
```
### before this PR
SGX=1 strace -e trace=read -s 4096 -ff ./pal_loader ./vfork_and_exec 2>&1 | tee strace_plain.log
# check strace_plain.log and closer to end you'll see a huge read() -- this is checkpoint
# in this read, you'll see plain strings like `pipe:3035966780` and `chroot`

### after this PR
SGX=1 strace -e trace=read -s 4096 -ff ./pal_loader ./vfork_and_exec 2>&1 | tee strace_enc.log
# check strace_enc.log and closer to end you'll see a huge read() -- this is checkpoint
# in this read, you'll see no discernible strings -- everything is encrypted
```


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/oscarlab/graphene/1229)
